src/workerd/api/streams/internal.c++

```diff
@@ -61,48 +61,46 @@ public:
   }
 
   kj::Promise<kj::Array<byte>> readAllBytes() {
-    return loop().then([this](PartList&& partPtrs) {
-      auto out = kj::heapArray<byte>(runningTotal);
-      copyInto(out, kj::mv(partPtrs));
-      return kj::mv(out);
-    });
+    auto parts = co_await readParts();
+    auto out = kj::heapArray<byte>(runningTotal);
+    copyInto(out, kj::mv(parts));
+    co_return out;
   }
 
   kj::Promise<kj::String> readAllText() {
-    return loop().then([this](PartList&& partPtrs) {
-      auto out = kj::heapArray<char>(runningTotal + 1);
-      copyInto(out.slice(0, out.size() - 1).asBytes(), kj::mv(partPtrs));
-      out.back() = '\0';
-      return kj::String(kj::mv(out));
-    });
+    auto parts = co_await readParts();
+    auto out = kj::heapArray<char>(runningTotal + 1);
+    copyInto(out.slice(0, out.size() - 1).asBytes(), kj::mv(parts));
+    out.back() = '\0';
+    co_return kj::String(kj::mv(out));
   }
 
 private:
   ReadableStreamSource& input;
   uint64_t limit;
-  kj::Vector<kj::Array<kj::byte>> parts;
   uint64_t runningTotal = 0;
 
-  kj::Promise<PartList> loop() {
-    auto bytes = kj::heapArray<kj::byte>(4096);
+  kj::Promise<PartList> readParts() {
+    static constexpr size_t bufferSize = 4096;
+    PartList parts;
+
+    while (true) {
+      auto part = kj::heapArray<kj::byte>(bufferSize);
+      auto amount = co_await input.tryRead(part.begin(), bufferSize, bufferSize);
+      runningTotal += amount;
+      JSG_REQUIRE(runningTotal < limit, TypeError, "Memory limit exceeded before EOF.");
 
-    return input.tryRead(bytes.begin(), 1, bytes.size())
-        .then([this, bytes = kj::mv(bytes)](size_t amount) mutable
-        -> kj::Promise<PartList> {
-      if (amount == 0) {
-        return KJ_MAP(p, parts) { return p.asPtr(); };
+      if (amount > 0) {
+        parts.add(part.slice(0, amount).attach(kj::mv(part)));
```

Comment:
Sorry to keep nitpicking here, but in the common case that `amount == bufferSize`, this `.attach()` is doing a redundant allocation.

Reply:
alright, I got rid of all intermediate copies here.